### PR TITLE
[DEV-1601] add user mails: pending, return mails

### PR DESCRIPTION
**Pull request by Yannick1712**
#### Pre-Release

- [ ] Check migrations
  - No database related infos (sqldb-xxx)
  - Impact on GS (new/removed columns)
- [ ] Check for linter errors (in PR)
- [ ] Test basic user operations
  - Login/logout
  - Registration
  - Update user data
  - Create/delete payment routes

#### Post-Release

- Test basic user operations
- Monitor application insights log

### CODE REVIEW

File: src/subdomains/supporting/notification/factories/mail.factory.ts [hunk 1]
```diff
@@ -256,17 +274,32 @@ export class MailFactory {
...
-    return match ? { text: match[1], tag: match[2], value: match[3] } : undefined;
+  private combineConnectedLineElements(suffix: TranslationItem[]): TranslationItem[] {
+    for (let index = 0; index < suffix.length - 1; index++) {
+      if (suffix[index]?.params?.connectNextLine) {
+        suffix[index + 1].key = `${suffix[index].key}&&${suffix[index + 1].key}`;
+        suffix[index] = null;
+      }
+    }
+    return suffix;
+  }
```
Comment (davidleomay) [inline src/subdomains/supporting/notification/factories/mail.factory.ts:299]: I wouldn't use this combine logic, but translation params. E.g.:
```ts
`{ key: `${MailTranslationKey.RETURN}.introduction`, params: { reason: `${Util.parseMailKey(MailTranslationKey.RETURN, entity.amlReason)}` } }`
```
and in the translation file
```json
"introduction": "Grund, warum wir zurückerstattet haben:<br>{reason}",
```
then translate
```ts
 this.tNew(element.key, lang, params);
```
where we should also translate all params.

File: src/subdomains/core/buy-crypto/process/services/buy-crypto-notification.service.ts [hunk 2]
```diff
@@ -65,34 +66,48 @@ export class BuyCryptoNotificationService {
...
+                    : null,
                 },
+                suffix: [
+                  { key: MailKey.SPACE, params: { value: '4' } },
+                  tx.isLightningOutput
+                    ? null
+                    : {
+                        key: `${MailTranslationKey.BUY_CRYPTO}.payment_link`,
+                        params: { url: txExplorerUrl(tx.target.asset.blockchain, tx.txId) },
+                      },
```
Comment (davidleomay) [inline src/subdomains/core/buy-crypto/process/services/buy-crypto-notification.service.ts:104]: The space above should only be added, if it is not a lightning output?
(also applies to all other mails)

File: src/subdomains/supporting/notification/factories/mail.factory.ts (hunk 1 above)
Comment (Yannick1712) [reply src/subdomains/supporting/notification/factories/mail.factory.ts:299]: But then we have again parameters in the translation json file 🤔 
Thats what we tried to undo with the Refactoring, or? 😅 
The nice thing with the Refactoring is that everyone can edit these files without any knowledge about parameters and mail logic and so on

File: src/subdomains/core/buy-crypto/process/entities/buy-crypto.entity.ts [hunk 3]
```diff
@@ -359,43 +359,28 @@ export class BuyCrypto extends IEntity {
...
+      (this.outputAmount / this.inputReferenceAmountMinusFee) * (this.inputReferenceAmount / this.inputAmount),
+      2,
+    )} ${this.outputAsset.name}/${this.inputAsset}`;
+  }
 
-    throw new Error(`Tried to send a mail for buy-crypto ${this.id} in invalid state`);
+  get translationKey(): MailTranslationKey {
+    if (!this.isCryptoCryptoTransaction) return MailTranslationKey.CRYPTO_RETURN;
+    return MailTranslationKey.FIAT_RETURN;
   }
```
[the remark this answers is not in the document]
Comment (Yannick1712) [reply src/subdomains/core/buy-crypto/process/entities/buy-crypto.entity.ts:384]: Hehe yeah we can rename it 👍

File: src/subdomains/core/buy-crypto/process/services/buy-crypto-notification.service.ts (hunk 2 above)
Comment (Yannick1712) [reply src/subdomains/core/buy-crypto/process/services/buy-crypto-notification.service.ts:104]: Oh yes good point 👍

File: src/subdomains/supporting/notification/factories/mail.factory.ts [hunk 4]
```diff
@@ -261,11 +268,8 @@ export class MailFactory {
...
       default:
         const text =
-          element.key.split('&&').length > 1
-            ? element.key
-                .split('&&')
-                .map((splitKey) => this.tNew(splitKey, lang))
-                .join('')
+          element.params && !this.isMailParamType(element.params)
+            ? this.getTranslatedTextWithParams(element, lang)
             : this.tNew(element.key, lang);
```
[the remark this answers is not in the document]
Comment (davidleomay) [reply src/subdomains/supporting/notification/factories/mail.factory.ts]: If you want you can additionally filter for mail param types after the map,

File: src/subdomains/supporting/notification/factories/mail.factory.ts [hunk 5]
```diff
@@ -292,19 +283,20 @@ export class MailFactory {
...
-  }
-
   private parseSpecialTag(text: string): SpecialTag | undefined {
     const match = /^(.*)\[(\w+):([^\]]+)\](.*)$/.exec(text);
     return match ? { text: match[1], textSuffix: match[4], tag: match[2], value: match[3] } : undefined;
   }
 
-  private isMailParamType(params: TranslationItem['params']): params is { [key in MailParamKey]?: string } {
-    return MailParamKey[Object.keys(params)[0]] !== undefined;
+  private translateParams(params: TranslationItem['params'], lang: string): TranslationItem['params'] {
```
Comment (davidleomay) [inline src/subdomains/supporting/notification/factories/mail.factory.ts:291]: Use a separate type for this:

```ts
export type TranslationParams = { [key in MailParamKey]?: string } | Record<string, string>;

export interface TranslationItem {
  key: string;
  params?: TranslationParams;
}
```

**Review by davidleomay — APPROVED** [review]

**Pull request merged**